Initial commit

zhanggbj opened the pull request:
This is an initial commit based on discussion https://github.com/knative/community/issues/211

zhanggbj commented:
@chizhg @mattmoor This is just an initial commit and more items will be tracked in the issues. Would you please help to take a review and merge it? Thanks!

chizhg commented on core/root.go:
```diff
@@ -0,0 +1,59 @@
+
+	rootCmd := &cobra.Command{
+		Use:   "kperf",
+		Short: "A CLI of to help with Knative performance test",
```
Remove `of`.

chizhg commented on core/root.go:
```diff
@@ -0,0 +1,59 @@
+		Long:  `A CLI of to help with Knative performance test.`,
+	}
+	cobra.OnInitialize(initConfig)
+	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.config/kubectl)")
```
The description here for defaulting is not the same as the actual logic, which one is correct?

zhanggbj replied:
This is not used. Removing it.

chizhg commented on go.mod:
```diff
@@ -0,0 +1,45 @@
+
+// Pinned to kubernetes-1.15.4
+replace (
+	k8s.io/api => k8s.io/api v0.0.0-20190918195907-bd6ac527cfd2
```
Better to pin to the same k8s version as other Knative repos, currently serving is pinning to v0.18.8 - https://github.com/knative/serving/blob/master/go.mod#L44

zhanggbj replied:
I haven't tested this yet, how about I open a separate task after this initial PR is merged and update it?
https://github.com/knative-sandbox/kperf/issues/6

chizhg commented on hack/build-flags.sh:
```diff
@@ -0,0 +1,15 @@
+  local base="${1}"
+  local now="$(date -u '+%Y-%m-%d %H:%M:%S')"
+  local rev="$(git rev-parse --short HEAD)"
+  local pkg="github.ibm.com/zhanggbj/kperf/pkg/command/version"
```
Please fix the pkg name.

chizhg commented on cmd/kperf.go:
```diff
@@ -0,0 +1,15 @@
+
+func main() {
+	if err := core.NewPerfCommand().Execute(); err != nil {
+		fmt.Println("failed to execute kperf command:", err)
```
Better to change all the `fmt.Print` to `log.Print`

zhanggbj replied:
Fixed

chizhg reviewed (commented):
This is a really huge PR. I left some comments but mostly are nits and general issues.

Before we move forward, I think a few issues need to be discussed and addressed:
1. Since the current kperf tool includes two parts, one part is for general benchmarking and charting, and the other is specific to Knative serving, I think productivity WG could be co-owning the first part and serving WG could be co-owning the second, @mattmoor does it make sense to you  ? If yes, @zhanggbj can we make the tool more modular by separating the code into packages with more meaningful names, and defining clearer ownerships?

2. The current implementation of `kperf` is specifically for benchmarking Knative serving control plane, how hard will it be to extend it to be able to benchmark Knative serving data plane? In other words, how much work will be needed to convert an existing benchmark like https://github.com/knative/serving/tree/master/test/performance/benchmarks/load-test/continuous to a `kperf` subcommand?

3. Is `kperf` designed to be able to support benchmarking other Knative projects like `eventing` as well? Or is it limited to `serving`?

Also according to our discussions in https://github.com/knative/community/issues/211, we were planning to use `kperf`'s charting support for local testing and debugging purpose, but still use Mako for benchmarking in the CI environment, so a few interfaces (in this case some flags) will need to be added to support streaming the data to Mako storage. After this PR is merged, I will start drafting a doc for `kperf`+`Mako` integration on how we can move forward with this path.

chizhg commented:
> This is a really huge PR. I left some comments but mostly are nits and general issues.
> 
> Before we move forward, I think a few issues need to be discussed and addressed:
> 
> 1. Since the current kperf tool includes two parts, one part is for general benchmarking and charting, and the other is specific to Knative serving, I think productivity WG could be co-owning the first part and serving WG could be co-owning the second, @mattmoor does it make sense to you  ? If yes, @zhanggbj can we make the tool more modular by separating the code into packages with more meaningful names, and defining clearer ownerships?
> 2. The current implementation of `kperf` is specifically for benchmarking Knative serving control plane, how hard will it be to extend it to be able to benchmark Knative serving data plane? In other words, how much work will be needed to convert an existing benchmark like https://github.com/knative/serving/tree/master/test/performance/benchmarks/load-test/continuous to a `kperf` subcommand?
> 3. Is `kperf` designed to be able to support benchmarking other Knative projects like `eventing` as well? Or is it limited to `serving`?
> 
> Also according to our discussions in [knative/community#211](https://github.com/knative/community/issues/211), we were planning to use `kperf`'s charting support for local testing and debugging purpose, but still use Mako for benchmarking in the CI environment, so a few interfaces (in this case some flags) will need to be added to support streaming the data to Mako storage. After this PR is merged, I will start drafting a doc for `kperf`+`Mako` integration on how we can move forward with this path.

cc @vagababov @julz @grantr @albertomilan for more input.

gyohuangxin commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,125 @@
+				end, _ := strconv.Atoi(r[1])
+				if start >= 0 && end >= 0 && start <= end {
+					for i := start; i <= end; i++ {
+						nsRangeMap[fmt.Sprintf("%s%d", nsPrefix, i)] = true
```
`nsRangeMap` in clean.go is %s%d, while in generate.go is %s-%d. They should be matched.

zhanggbj replied:
@gyohuangxin nice catch, thanks! This will be refined in https://github.com/knative-sandbox/kperf/issues/4 after the initial PR merged.

vagababov replied:
Why not fix it right here?

gyohuangxin commented on pkg/command/service/generate.go:
```diff
@@ -0,0 +1,186 @@
+				end, _ := strconv.Atoi(r[1])
+				if start > 0 && end > 0 && start <= end {
+					for i := start; i <= end; i++ {
+						nsRangeMap[fmt.Sprintf("%s-%d", nsPrefix, i)] = true
```
As mentioned above, it should match the `nsRangeMap` in clean.go.

zhanggbj replied:
@gyohuangxin nice catch, thanks! This will be refined in https://github.com/knative-sandbox/kperf/issues/4 after the initial PR merged.

gyohuangxin replied:
Get it, thanks!

zhanggbj commented on OWNERS:
```diff
@@ -0,0 +1,8 @@
+- maximilien
+# Also including TOC members as backup:
+- mattmoor
+- chizhg 
```
@chizhg Sure, will update in the next patch!

zhanggbj commented on README.md:
```diff
@@ -1 +1,149 @@
+Please **NOTE** this project is under development.
+Kperf is designed for Knative Load test. It helps to generate workload like Knative services and 
+gives measurement result about underneath resource create time duration based on server side timestamps, 
+and give statics and raw measurement result, to help Knative developers or operators to figure out Kntive platform
```
Fixed!

zhanggbj commented on cmd/kperf.go:
```diff
@@ -0,0 +1,15 @@
+	"fmt"
+	"os"
+
+	"github.com/zhanggbj/kperf/core"
```
Fixed!

zhanggbj commented on templates/single_chart.html:
```diff
@@ -0,0 +1,334 @@
+                    formatter: function (params) {
+                        var content = params[0].axisValue + '<br/>'
+                        params.forEach(item => {
+                            var icon = '<svg viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" width="12" height="12"><path stroke="' + item.color + '" fill="' + item.color + '" d="' + legendIcons[legendGroupMap[legendMap[item.seriesName]]].replace('path://', '') + '"/></g></svg>'
```
Fixed!

zhanggbj commented on README.md:
```diff
@@ -1 +1,149 @@
+
+## User Stories
+
+### As a Codeengine developer, I want to generate Codeengine Service concurrently for test
```
Fixed

zhanggbj commented on cmd/kperf.go:
```diff
@@ -0,0 +1,15 @@
+package main
```
Fixed!

zhanggbj commented on README.md:
````diff
@@ -1 +1,149 @@
+## Build kperf
+```cassandraql
+# format and build kperf
+$ cd {workspace}/src/github.com/zhanggbj/kperf
````
Fixed

zhanggbj commented:
Thanks a lot for your review @chizhg @gyohuangxin. I've addressed the comments and please help to take look again.
As this is an initial PR, some known item like UT will be addressed in separate PR which is more easier to review. Thanks!

zhanggbj commented:
@chizhg About the questions,

> Before we move forward, I think a few issues need to be discussed and addressed:
> 
> 1. Since the current kperf tool includes two parts, one part is for general benchmarking and charting, and the other is specific to Knative serving, I think productivity WG could be co-owning the first part and serving WG could be co-owning the second, @mattmoor does it make sense to you  ? If yes, @zhanggbj can we make the tool more modular by separating the code into packages with more meaningful names, and defining clearer ownerships?
> 2. The current implementation of `kperf` is specifically for benchmarking Knative serving control plane, how hard will it be to extend it to be able to benchmark Knative serving data plane? In other words, how much work will be needed to convert an existing benchmark like https://github.com/knative/serving/tree/master/test/performance/benchmarks/load-test/continuous to a `kperf` subcommand?
> 3. Is `kperf` designed to be able to support benchmarking other Knative projects like `eventing` as well? Or is it limited to `serving`?
> 
> Also according to our discussions in [knative/community#211](https://github.com/knative/community/issues/211), we were planning to use `kperf`'s charting support for local testing and debugging purpose, but still use Mako for benchmarking in the CI environment, so a few interfaces (in this case some flags) will need to be added to support streaming the data to Mako storage. After this PR is merged, I will start drafting a doc for `kperf`+`Mako` integration on how we can move forward with this path.

For 1) About ownership, let's wait for @mattmoor 's comments.
Personally, I prefer to keep the current directory, as it's more clear to manage them based on scenarios like service or eventing sources, but not by ownership/

For 2) In the initial PR, there's only benchmarking for Knative service deployment (control plane). We already have benchmarking for Knative service access (Data plane). If needed we can give a brief introduction/demo about it.
However I think we can firstly make the whole process work, taking control plane as an example and then add benchmarking for Data plane

For 3) I think kperf is not just limited to Serving. We can easily extend it to support Other Knative Benchmarking if needed, like eventing related.

mattmoor commented:
Without looking...  I'd expect productivity to own the common infrastructure for facilitating benchmarking and downstream WGs to own the specific tie-ins (e.g. serving + control plane, networking + dataplane, autoscaling + cold start).

vagababov commented on cmd/kperf.go:
```diff
@@ -0,0 +1,29 @@
+// Copyright © 2020 The Knative Authors
```
I think our copyright header is (c), feels a bit off :)

zhanggbj replied:
@vagababov Nice catch. Just want to double confirm, should I follow (C)
or withour C like kn clien did `Copyright 2020 The Knative Authors`
https://github.com/knative/client/blob/master/pkg/apis/client/register.go#L2

chizhg replied:
I think without C, it's also what we are using for serving.

vagababov commented on core/root.go:
```diff
@@ -0,0 +1,72 @@
+	viper.AutomaticEnv() // read in environment variables that match
+
+	// If a config file is found, read it in.
+	if err := viper.ReadInConfig(); err == nil {
```
what if it fails?

zhanggbj replied:
@vagababov I think this part is generated code by cobra
it will fail with err
```
2020/10/29 11:45:00 failed to execute kperf command: invalid configuration: no configuration has been provided
```

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+			var nsRangeMap map[string]bool = map[string]bool{}
+			if nsPrefix != "" {
+				r := strings.Split(nsRange, ",")
+				if len(r) != 2 {
+					fmt.Printf("Expected Range like 1,500, given %s\n", nsRange)
+					os.Exit(1)
```
this is basically `log.Fatalf`.

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+					fmt.Printf("Expected Range like 1,500, given %s\n", nsRange)
+					os.Exit(1)
+				}
+				start, _ := strconv.Atoi(r[0])
+				end, _ := strconv.Atoi(r[1])
```
don't ignore errors.
Also you can use `fmt.Sscanf` to read the whole pair.

zhanggbj replied:
@vagababov Thanks! Will fix it in next patch.

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+				}
+				start, _ := strconv.Atoi(r[0])
+				end, _ := strconv.Atoi(r[1])
+				if start >= 0 && end >= 0 && start <= end {
```
If the input is negative, there should be an error as well.

zhanggbj replied:
Will fix it in next patch!

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+kperf service clean --nsprefix testns/ --ns nsname
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var nsRangeMap map[string]bool = map[string]bool{}
```
just use `sets.String`

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+					return fmt.Errorf("no namespace found with prefix %s", nsPrefix)
+				}
+			} else {
+				return fmt.Errorf("both ns and nsPrefix are empty")
```
there's no formatting.
Use `errors.New()`.

zhanggbj replied:
Fixed!

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+			if len(matchedNsNameList) > 0 {
+				generator.NewBatchCleaner(matchedNsNameList, concurrency, cleanKsvc).Clean()
+			} else {
+				fmt.Printf("No service found for cleaning\n")
```
```suggestion
				fmt.Println("No service found for cleaning")
```

zhanggbj replied:
Fixed!

vagababov commented on pkg/command/service/clean.go:
```diff
@@ -0,0 +1,139 @@
+			} else {
+				return fmt.Errorf("both ns and nsPrefix are empty")
+			}
+			matchedNsNameList := [][2]string{}
+			for i := 0; i < len(nsNameList); i++ {
+				svcList, err := ksvcClient.Services(nsNameList[i]).List(metav1.ListOptions{})
+				if err == nil {
+					for j := 0; j < len(svcList.Items); j++ {
+						if strings.HasPrefix(svcList.Items[j].Name, svcPrefix) {
+							matchedNsNameList = append(matchedNsNameList, [2]string{nsNameList[i], svcList.Items[j].Name})
+						}
+					}
+				}
+			}
```
This and file in general needs much more comments.

zhanggbj replied:
Totally make sense, will cover in https://github.com/knative-sandbox/kperf/issues/4, as @qibobo is working another PR to refine them.

vagababov commented on pkg/command/service/generate.go:
```diff
@@ -0,0 +1,200 @@
+	"knative.dev/kperf/pkg/generator"
+)
+
+var (
+	count, interval, batch, concurrency, minScale, maxScale int
+	nsPrefix, nsRange, ns                                   string
+	svcNamePrefixDefault                                    string = "testksvc"
+	checkReady                                              bool
+	timeout                                                 time.Duration
+	ksvcClient                                              *servingv1client.ServingV1Client
+	err                                                     error
+)
```
Do we need global variables?

zhanggbj replied:
This part looks ugly to me and we're refining all the variable definitions, covered in #4. Will update later. Thanks!

vagababov commented on pkg/command/service/generate.go:
```diff
@@ -0,0 +1,200 @@
+			if nsPrefix != "" {
+				r := strings.Split(nsRange, ",")
+				if len(r) != 2 {
+					fmt.Printf("Expected Range like 1,500, given %s\n", nsRange)
+					os.Exit(1)
```
Ditto same comments in this file. Although it looks like it's the same code, so you need to extract this to a library, I think.

zhanggbj replied:
Make sense to me. Will add a common function for parsing, covered #4

vagababov commented on pkg/command/service/generate.go:
```diff
@@ -0,0 +1,200 @@
+		Spec: servingv1.RevisionSpec{},
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"autoscaling.knative.dev/minScale": fmt.Sprintf("%d", minScale),
+				"autoscaling.knative.dev/maxScale": fmt.Sprintf("%d", maxScale),
```
```suggestion
				"autoscaling.knative.dev/minScale": strconv.Itoa(minScale),
				"autoscaling.knative.dev/maxScale": strconv.Itoa(maxScale),
```

zhanggbj replied:
Fixed!

vagababov commented on pkg/command/service/generate.go:
```diff
@@ -0,0 +1,200 @@
+		},
+	}
+
+	service.Spec.Template = servingv1.RevisionTemplateSpec{
+		Spec: servingv1.RevisionSpec{},
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"autoscaling.knative.dev/minScale": fmt.Sprintf("%d", minScale),
+				"autoscaling.knative.dev/maxScale": fmt.Sprintf("%d", maxScale),
+			},
+		},
+	}
+	service.Spec.Template.Spec.Containers = []corev1.Container{
+		{
+			Image: "docker.io/qibobo/go4autoscaler:latest",
+			Ports: []corev1.ContainerPort{
+				{
+					ContainerPort: 8080,
+				},
+			},
+		},
+	}
```
Those should be kind of joined together.
Also we should not check in this using some random image.

zhanggbj replied:
How about switch to `gcr.io/knative-samples/knative-route-demo` from knative community doc?

zhanggbj replied:
For now I updated to "gcr.io/knative-samples/helloworld-go".

vagababov commented on pkg/command/service/generate.go:
```diff
@@ -0,0 +1,200 @@
+	fmt.Printf("Creating ksvc %s in namespace %s\n", service.GetName(), service.GetNamespace())
+	_, err := ksvcClient.Services(ns).Create(&service)
+	if err != nil {
+		fmt.Printf("Failed to create ksvc %s in namespace %s : %s\n", service.GetName(), service.GetNamespace(), err)
```
Should the system exit here?

zhanggbj replied:
As we're generating service creation load, I don't think one failure should stop the generation.
Measurement will finally calculate service creation suc/err rate.

zhanggbj commented:
Hi @vagababov @chizhg @gyohuangxin,

Thanks a lot for your review!  I've addressed all your comments mostly in this PR, some of the comments are tracking in `issues`, and other folks already are working on them.

Would you please help to review and merge it? I know there're a lot needs to be refined, but as it is an initial PR and is blocking all the other items. I suggest we merge it firstly,  so that others can continue work on the items listed in issues.

Thanks!
CC @maximilien

maximilien commented on hack/build.sh:
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
```
Add copyright

maximilien commented on pkg/generator/batch_generator.go:
```diff
@@ -0,0 +1,107 @@
+			ns := bg.namespaceList[index%len(bg.namespaceList)]
+			ns, name := bg.generateFunc(ns, index)
+			if bg.postGeneratorFunc(ns, name) != nil {
+				os.Exit(1)
```
```suggestion
				os.Exit(1)
```
Should we put some log or message to user here?

maximilien reviewed (approved):
Some minor updates / comments. Also please add some tests (at least e2e or both unit and e2e). Or if you want to add as a separate PR then create issue and link here. 

Otherwise /lgtm

chizhg commented:
/approve

Please address the following comments in followup PRs:
https://github.com/knative-sandbox/kperf/pull/3/files#r509880726
https://github.com/knative-sandbox/kperf/pull/3/files#r518227195
https://github.com/knative-sandbox/kperf/pull/3/files#r509883984
https://github.com/knative-sandbox/kperf/pull/3/files#r509881423

And also I think all the `fmt.Print` occurrences should be replaced with `log.Print` or `log.Fatal` if followed with `os.Exit(1)`

chizhg commented:
/lgtm

chizhg commented:
/approve
